Complete outstanding tasks for 1.21.3 update

adamkahl opened the pull request:
This is my first PR for this repository, and I wanted to help out with the update to 1.21.3, so I completed the tasks listed in the PR for minecraft-data https://github.com/PrismarineJS/minecraft-data/pull/936.  
I obtained most of the data from [wiki.vg](https://wiki.vg/Pre-release_protocol), and some from [Extracted Minecraft Data](https://github.com/extremeheat/extracted_minecraft_data/tree/client1.21.3).

The updates from the PR task list include:
- packet_recipe_book_add
- packet_craft_recipe_response
- packet_declare_recipes

And a few other updates due to tests not passing (following the changes in the wiki):
- MovementFlags
- packet_steer_vehicle
- packet_position

A couple of other notes:
- I also switched to the bitfields version of PositionUpdateRelatives to get the tests to pass. 
- I did not update anything in packet_set_cooldown, because the wiki did not show any changes to the protocol. 

If there is more work to be done, I would be happy to help out, but would need some guidance for where to contribute.

adamkahl committed:
Updated packet_craft_recipe_response, packet_recipe_book_add, packet_declare_recipes, MovementFlags, and packet_steer_vehicle

GroobleDierne commented on data/pc/latest/proto.yml:
```diff
@@ -2120,23 +2120,24 @@
          default: void
 
    # https://github.com/extremeheat/extracted_minecraft_data/blob/client1.21.2/client/net/minecraft/world/entity/Relative.java#L82
-   PositionUpdateRelatives: ["bitflags", {
```
What is the reason to replace every bitflags with bitfields? You said it wasn't passing test but do you know why they failed?

adamkahl replied:
Here's what I was running in to:
`{
  "keyword": "enum",
  "dataPath": "",
  "schemaPath": "#/oneOf/0/enum",
  "params": {
    "allowedValues": [
      "native"
    ]
  },
  "message": "should be equal to one of the allowed values",
  "schema": [
    "native"
  ],
  "parentSchema": {
    "enum": [
      "native"
    ]
  },
  "data": [
    "bitflags",
    {
      "type": "u32",
      "flags": [
        "x",
        "y",
        "z",
        "yaw",
        "pitch",
        "dx",
        "dy",
        "dz",
        "yawDelta"
      ]
    }
  ]
}`
I could not figure out why the bitflags was not working but the bitfields was. I don't think i'm quite familiar enough with the validation to debug this.

GroobleDierne replied:
Oh that's because bitflag is a new type defined [here](https://github.com/PrismarineJS/node-minecraft-protocol/compare/master...1.21.3)

GroobleDierne commented on data/pc/latest/proto.yml:
```diff
@@ -2623,16 +2623,18 @@
    packet_declare_recipes:
-      recipes: []varint
-         name: string
+      propertySetCount: varint
```
why? `properySets: []varint` already reads the length of the array

adamkahl replied:
I was just following the protocol shown [here](https://wiki.vg/Pre-release_protocol#Update_Recipes), if its not needed then `propertySetCount` could probably be removed.

GroobleDierne replied:
In `properySets: []varint` the varint means that the size of this array is given by a varint field preceding the array.

GroobleDierne reviewed (commented):
I'll test the changes properly in a few hours.

adamkahl commented:
Are there more resources for how to properly test this repository for a new version? I'm happy to help out with the process but I could not find any resources in the documentation for how complete testing is done.

GroobleDierne commented on data/pc/latest/proto.yml:
```diff
@@ -2155,71 +2156,70 @@
+         - stonecutter
+         - smithing
+         - campfire
+         hasIngredients: bool
```
This is unneeded, `craftingRequirements?:` does the same, `?` means it's an optional field

GroobleDierne commented:
Considering the issues I brought up, I don't think I'll be merging as it is basically the same protocol with a more verbose specification. Feel free to point out any real protocol changes I missed when reviewing

GroobleDierne commented:
> Are there more resources for how to properly test this repository for a new version? I'm happy to help out with the process but I could not find any resources in the documentation for how complete testing is done.

Not really, what I do is, I clone NMP, run `npm i`, replace the files I want in node_modules/minecraft-data/minecraft-data/data/ (If you are adding a brand new version you'll have to run `npm run generate:data` the first time) and then I use the client_chat example in NMP to connect to a vanilla server.

adamkahl commented on data/pc/latest/proto.yml:
```diff
@@ -3141,11 +3145,16 @@
-      ## client/net/minecraft/world/entity/player/Input.java
-      inputs: ["bitflags", {
-         "type": "u8",
-         "flags": ["forward", "backward", "left", "right", "jump", "sneak", "sprint"]
```
This is changed from `sneak` to `shift`

adamkahl commented:
> Considering the issues I brought up, I don't think I'll be merging as it is basically the same protocol with a more verbose specification. Feel free to point out any real protocol changes I missed when reviewing

I added a 2 comments of real protocol changes I updated.
